File: VB/RuntimeDbChooser.Win/WinApplicationEx.vb
Imports System
Imports System.Configuration
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Win
Imports DevExpress.ExpressApp.Security
Imports RuntimeDbChooser.Module.Win
Imports RuntimeDbChooser.Module.BusinessObjects

Namespace RuntimeDbChooser.Win
    Partial Public Class RuntimeDbChooserWindowsFormsApplication
        Inherits WinApplication
        Implements IApplicationFactory

        Protected Overrides Sub ReadLastLogonParametersCore(ByVal storage As DevExpress.ExpressApp.Utils.SettingsStorage, ByVal logonObject As Object)
            Dim standardLogonParameters As AuthenticationStandardLogonParameters = TryCast(logonObject, AuthenticationStandardLogonParameters)
            If (standardLogonParameters IsNot Nothing) AndAlso String.IsNullOrEmpty(standardLogonParameters.UserName) Then
                MyBase.ReadLastLogonParametersCore(storage, logonObject)
            End If
        End Sub
        Protected Overrides Sub OnLoggingOn(ByVal args As LogonEventArgs)
            MyBase.OnLoggingOn(args)
            MSSqlServerChangeDatabaseHelper.UpdateDatabaseName(Me, DirectCast(args.LogonParameters, IDatabaseNameParameter).DatabaseName)
        End Sub
        Protected Overrides Function OnLogonFailed(ByVal logonParameters As Object, ByVal e As Exception) As Boolean
            If WinChangeDatabaseHelper.SkipLogonDialog Then
                Return True
            End If
            Return MyBase.OnLogonFailed(logonParameters, e)
        End Function
        Private Function IApplicationFactory_CreateApplication() As WinApplication Implements IApplicationFactory.CreateApplication
            Return CreateApplication()
        End Function
        Public Shared Function CreateApplication() As RuntimeDbChooserWindowsFormsApplication
            Dim winApplication As New RuntimeDbChooserWindowsFormsApplication()

            CType(winApplication.Security, SecurityStrategyComplex).Authentication = New WinChangeDatabaseStandardAuthentication()

            'WinChangeDatabaseActiveDirectoryAuthentication activeDirectoryAuthentication = new WinChangeDatabaseActiveDirectoryAuthentication();
            'activeDirectoryAuthentication.CreateUserAutomatically = true;
            '((SecurityStrategyComplex)winApplication.Security).Authentication = activeDirectoryAuthentication;

            If ConfigurationManager.ConnectionStrings("ConnectionString") IsNot Nothing Then
                winApplication.ConnectionString = ConfigurationManager.ConnectionStrings("ConnectionString").ConnectionString
            End If
            Return winApplication
        End Function
    End Class
End Namespace

File: CS/RuntimeDbChooser.Win/WinApplicationEx.cs
using System;
using System.Configuration;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Win;
using DevExpress.ExpressApp.Security;
using RuntimeDbChooser.Module.Win;
using RuntimeDbChooser.Module.BusinessObjects;

namespace RuntimeDbChooser.Win {
	public partial class RuntimeDbChooserWindowsFormsApplication : WinApplication, IApplicationFactory {
		protected override void ReadLastLogonParametersCore(DevExpress.ExpressApp.Utils.SettingsStorage storage, object logonObject) {
			AuthenticationStandardLogonParameters standardLogonParameters = logonObject as AuthenticationStandardLogonParameters;
			if((standardLogonParameters != null) && string.IsNullOrEmpty(standardLogonParameters.UserName)) {
				base.ReadLastLogonParametersCore(storage, logonObject);
			}
		}
		protected override void OnLoggingOn(LogonEventArgs args) {
			base.OnLoggingOn(args);
			MSSqlServerChangeDatabaseHelper.UpdateDatabaseName(this, ((IDatabaseNameParameter)args.LogonParameters).DatabaseName);
		}
		protected override bool OnLogonFailed(object logonParameters, Exception e) {
			if(WinChangeDatabaseHelper.SkipLogonDialog) {
				return true;
			}
			return base.OnLogonFailed(logonParameters, e);
		}
		WinApplication IApplicationFactory.CreateApplication() {
			return CreateApplication();
		}
		public static RuntimeDbChooserWindowsFormsApplication CreateApplication() {
			RuntimeDbChooserWindowsFormsApplication winApplication = new RuntimeDbChooserWindowsFormsApplication();

			((SecurityStrategyComplex)winApplication.Security).Authentication = new WinChangeDatabaseStandardAuthentication();

			//WinChangeDatabaseActiveDirectoryAuthentication activeDirectoryAuthentication = new WinChangeDatabaseActiveDirectoryAuthentication();
			//activeDirectoryAuthentication.CreateUserAutomatically = true;
			//((SecurityStrategyComplex)winApplication.Security).Authentication = activeDirectoryAuthentication;

			if(ConfigurationManager.ConnectionStrings["ConnectionString"] != null) {
				winApplication.ConnectionString = ConfigurationManager.ConnectionStrings["ConnectionString"].ConnectionString;
			}
			return winApplication;
		}
	}
}
